Extend `RpoFalcon512ProcedureAcl` to note-based tx introspection

partylikeits1983 opened the pull request:
This PR resolves #1544 

This PR does the following:
* Updates `auth__tx_rpo_falcon512_procedure_acl` to have note based introspection, i.e. check if a note was created, and if so check if this is allowed by reading the flag from storage `allow_unauthorized_output_notes`.
* Adds tests for the `auth__tx_rpo_falcon512_procedure_acl` procedure (these tests may be removed if they seem redundant to the existing `test_rpo_falcon_procedure_acl` test).

I modified the `NUM_AUTH_TRIGGER_PROCS_SLOT` so that the storage layout is the following in MASM (in Rust it is reversed): 
```
[num_tracked_procs, allow_unauthorized_output_notes, 0, 0]
```

This is now the logic that checks for signatures in the `auth__tx_rpo_falcon512_procedure_acl` procedure:
```
if were_output_notes_created && !allow_unauthorized_output_notes
   require signature
else if was_tracked_proc_called
   require signature
else
    increment nonce
end
```

Comments:

1. Some of these stack variable names can be shorted. Additionally, I think we should put in the comment what "ACL" means, I still am unsure what this means, is ACL "Access Control List"?

2. I am using `loc_store` & `loc_load` for simplicity in accessing `allow_unauthorized_output_notes` after the loop.

mmagician commented on crates/miden-lib/src/account/auth/mod.rs:
```diff
@@ -58,24 +58,27 @@ impl From<RpoFalcon512> for AccountComponent {
 /// The storage layout is:
 /// - Slot 0(value): Public key (same as RpoFalcon512)
-/// - Slot 1(value): Number of trigger procedures
+/// - Slot 1(value): [0, 0, allow_unauthorized_output_notes, num_tracked_procs]
```
Oh nice, thanks for making this explicitly into a `WORD`, I realize now "Number of trigger procedures" wasn't really hinting at this being a full `WORD` (even if it's just one field element in practice) 👍🏼

mmagician commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -26,9 +26,15 @@ const.AUTH_TRIGGER_PROCS_MAP_SLOT=2
+export.auth__tx_rpo_falcon512_procedure_acl.1
     # Get the number of trigger procedures
-    push.NUM_AUTH_TRIGGER_PROCS_SLOT exec.account::get_item drop drop drop
+    push.NUM_AUTH_TRIGGER_PROCS_SLOT exec.account::get_item
+    # => [num_auth_trigger_procs, allow_unauthorized_output_notes, 0, 0, pad(16)]
+
+    movup.2 movup.3 drop drop
+    # => [num_auth_trigger_procs, allow_unauthorized_output_notes, pad(16)]
+
+    swap loc_store.0
```
If you keep the previous order, then you'd save three instructions here:
```
push.NUM_AUTH_TRIGGER_PROCS_SLOT exec.account::get_item
# => [0, 0, allow_unauthorized_output_notes, num_auth_trigger_procs, pad(16)]

drop drop
# => [allow_unauthorized_output_notes, num_auth_trigger_procs, pad(16)]

loc_store.0
```

mmagician commented on crates/miden-lib/src/account/auth/mod.rs:
```diff
@@ -142,8 +159,10 @@ mod tests {
-        let num_procs_slot = account.storage().get_item(1).expect("storage slot 1 access failed");
-        assert_eq!(num_procs_slot, Word::empty());
+        let slot_1 = account.storage().get_item(1).expect("storage slot 1 access failed");
+        // Slot 1 should be [0, 0, allow_unauthorized_output_notes, num_tracked_procs]
+        // With 0 procedures and allow_unauthorized_output_notes=false, this should be [0, 0, 0, 0]
+        assert_eq!(slot_1, Word::empty());
```
thanks, that's a good comment addition

mmagician commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -26,9 +26,15 @@ const.AUTH_TRIGGER_PROCS_MAP_SLOT=2
 #! Inputs:  [pad(16)]
 #! Outputs: [pad(16)]
-export.auth__tx_rpo_falcon512_procedure_acl
+export.auth__tx_rpo_falcon512_procedure_acl.1
```
I would also update the documentation above of this procedure

mmagician commented on crates/miden-lib/src/account/auth/mod.rs:
```diff
@@ -58,24 +58,27 @@ impl From<RpoFalcon512> for AccountComponent {
 pub struct RpoFalcon512ProcedureAcl {
     public_key: PublicKey,
     auth_trigger_procedures: Vec<Word>,
+    allow_unauthorized_output_notes: bool,
```
We should explain what is meant by `allow_unauthorized_output_notes` (ideally in the documentation of this struct).
In practice this means we disallow note anonymization / mixing, but as it's quite a "heavy" word, so I'd avoid referring to mixing if possible.

mmagician commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -62,7 +68,27 @@ export.auth__tx_rpo_falcon512_procedure_acl
+    and
+    # => [require_auth, was_any_called, i-1, pad(16)]
+
+    if.true
+        exec.::miden::contracts::auth::basic::auth__tx_rpo_falcon512
+    end
```
This would result in verifying the signature twice (it happens again below in case trigger procedures were called)

bobbinth commented:
> [num_tracked_procs, allow_unauthorized_output_notes, 0, 0]

I thought we also wanted to have a check of unauthorized input notes (see https://github.com/0xMiden/miden-base/issues/1544#issuecomment-3055717481).

PhilippGackstatter commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -16,19 +16,27 @@ const.NUM_AUTH_TRIGGER_PROCS_SLOT=1
-#! transaction. If none were called, authentication is skipped. If at least one was called,
-#! the standard RpoFalcon512 signature verification is performed.
-#! Nonce is incremented in both cases.
+#! This authentication procedure checks:
+#! 1. If any of the trigger procedures were called during the transaction
+#! 2. If input notes were consumed and allow_unauthorized_input_notes is false
+#! 3. If output notes were created and allow_unauthorized_output_notes is false
+#!
+#! If any of these conditions are true, standard RpoFalcon512 signature verification is performed.
+#! Otherwise, only the nonce is incremented.
```
Great explanation - easy to understand!

PhilippGackstatter commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -62,10 +70,44 @@ export.auth__tx_rpo_falcon512_procedure_acl
+    exec.tx::get_output_notes_commitment
+    # => [OUTPUT_NOTES_COMMITMENT, was_any_called, i-1, pad(16)]
+
+    padw eqw push.0 eq movdn.8 dropw dropw
+    # => [were_output_notes_created, was_any_called, i-1, pad(16)]
```
Nit: The result of `eqw` basically says `!were_output_notes_created` so instead of `push.0 eq` I'd just use a `not` to negate it to get `were_output_notes_created`.

PhilippGackstatter commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -62,10 +70,44 @@ export.auth__tx_rpo_falcon512_procedure_acl
+    exec.tx::get_input_notes_commitment
+    # => [INPUT_NOTES_COMMITMENT, require_output_note_auth, was_any_called, i-1, pad(16)]
+
+    padw eqw push.0 eq movdn.8 dropw dropw
```
Similar comment here regarding replacing `push.0 eq` with `not`.

PhilippGackstatter commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -62,10 +70,44 @@ export.auth__tx_rpo_falcon512_procedure_acl
+    # ------ Check if output notes were created ------
+
+    exec.tx::get_output_notes_commitment
+    # => [OUTPUT_NOTES_COMMITMENT, was_any_called, i-1, pad(16)]
```
Nit: I don't think we need `i-1` anymore after the loop, right? So should we just drop it here to avoid having to pass it through in every stack state comment?

PhilippGackstatter commented on crates/miden-lib/build.rs:
```diff
@@ -88,7 +88,8 @@ fn main() -> Result<()> {
     // compile transaction kernel
     let mut assembler =
-        compile_tx_kernel(&source_dir.join(ASM_TX_KERNEL_DIR), &target_dir.join("kernels"))?;
+        compile_tx_kernel(&source_dir.join(ASM_TX_KERNEL_DIR), &target_dir.join("kernels"))?
+            .with_debug_mode(true);
```
Nit: Should probably be reverted?

PhilippGackstatter commented on crates/miden-lib/src/account/auth/mod.rs:
```diff
@@ -46,36 +46,72 @@ impl From<RpoFalcon512> for AccountComponent {
 ///
-/// The storage layout is:
+/// Authentication is required if ANY of the following conditions are true:
+/// - Any trigger procedure from the ACL was called
```
As you mentioned elsewhere, should we introduce the "ACL" acronym somewhere close to the beginning of this doc section, e.g. "Access Control List (ACL)"

PhilippGackstatter commented on crates/miden-lib/src/account/auth/mod.rs:
```diff
@@ -95,9 +136,25 @@ impl From<RpoFalcon512ProcedureAcl> for AccountComponent {
+        // allow_unauthorized_input_notes, 0]
         let num_procs = falcon.auth_trigger_procedures.len() as u32;
-        storage_slots.push(StorageSlot::Value(Word::from([num_procs, 0, 0, 0])));
+        let allow_unauthorized_output_notes = if falcon.allow_unauthorized_output_notes {
+            1u32
+        } else {
+            0u32
+        };
+        let allow_unauthorized_input_notes = if falcon.allow_unauthorized_input_notes {
+            1u32
+        } else {
+            0u32
+        };
+        storage_slots.push(StorageSlot::Value(Word::from([
+            num_procs,
+            allow_unauthorized_output_notes,
+            allow_unauthorized_input_notes,
+            0,
+        ])));
```
Nit: There is a conversion from [`bool` to `u32`](https://doc.rust-lang.org/std/primitive.bool.html#impl-From%3Cbool%3E-for-u32), so this could be written a bit more concisely (using `From` or `as`):

```rust
storage_slots.push(StorageSlot::Value(Word::from([
    num_procs,
    u32::from(falcon.allow_unauthorized_output_notes),
    u32::from(falcon.allow_unauthorized_input_notes),
    0,
])));
```

PhilippGackstatter commented on crates/miden-lib/src/account/auth/mod.rs:
```diff
@@ -130,8 +187,8 @@ mod tests {
         let public_key = PublicKey::new(Word::empty());
-        let component =
-            RpoFalcon512ProcedureAcl::new(public_key, vec![]).expect("component creation failed");
+        let component = RpoFalcon512ProcedureAcl::new(public_key, vec![], false, false)
+            .expect("component creation failed");
```
Nit: I think using `bool`s here is suboptimal, because `false, false` here doesn't tell the reader much about what this is doing, and having two bools next to each other seems easy to mix up.

This would be a cool usecase for [`bitflags`](https://docs.rs/bitflags/latest/bitflags/), e.g.:

```rust

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct NoteAuthorizationFlags(u8);

bitflags::bitflags! {
    impl NoteAuthorizationFlags: u8 {
        const UnauthorizedInputNotes = 0b00000001;
        const UnauthorizedOutputNotes = 0b00000010;
    }
}
```

And then the constructor would be very readable:

```rust
let component = RpoFalcon512ProcedureAcl::new(
    public_key,
    vec![],
    NoteAuthorizationFlags::UnauthorizedInputNotes | NoteAuthorizationFlags::UnauthorizedOutputNotes,
)
```

But not sure it's worth adding the dependency for.

PhilippGackstatter commented on crates/miden-lib/src/account/auth/mod.rs:
```diff
@@ -142,8 +199,12 @@ mod tests {
-        let num_procs_slot = account.storage().get_item(1).expect("storage slot 1 access failed");
-        assert_eq!(num_procs_slot, Word::empty());
+        let slot_1 = account.storage().get_item(1).expect("storage slot 1 access failed");
+        // Slot 1 should be [num_tracked_procs, allow_unauthorized_output_notes,
+        // allow_unauthorized_input_notes, 0] With 0 procedures,
+        // allow_unauthorized_output_notes=false, and allow_unauthorized_input_notes=false, this
+        // should be [0, 0, 0, 0]
+        assert_eq!(slot_1, Word::empty());
```
Nit: Should we put the sentence starting with "With 0 procedures" on its own line for readability?

PhilippGackstatter commented on crates/miden-lib/src/account/faucets/mod.rs:
```diff
@@ -246,7 +247,7 @@ pub fn create_basic_fungible_faucet(
 
     let auth_component: RpoFalcon512ProcedureAcl = match auth_scheme {
         AuthScheme::RpoFalcon512 { pub_key } => {
-            RpoFalcon512ProcedureAcl::new(pub_key, vec![distribute_proc_root])
+            RpoFalcon512ProcedureAcl::new(pub_key, vec![distribute_proc_root], false, false)
                 .map_err(FungibleFaucetError::AccountError)?
         },
     };
```
This is part of `create_basic_fungible_faucet`, so should we now change this to:

```rust
RpoFalcon512ProcedureAcl::new(pub_key, vec![], false, true)
```

Which would allow anyone to burn assets by executing transactions against this account but creating assets and creating notes would require authentication.

cc @mmagician 

Even if we don't do this, at least we have to `allow_unauthorized_input_notes = true`, because otherwise the `burn` functionality is inaccessible I think?

mmagician replied:
> The above shows that using bools makes it impossible to figure out which one applies to input and output notes when talking about a piece of code in isolation.

Agreed, were you thinking of enums for this?

PhilippGackstatter replied:
> Agreed, were you thinking of enums for this?

That's possible, but not sure they're well-suited for something like this. Just to showcase, using enums, we could do it one of two ways I think:

**Option 1**

```rust
enum NoteAuthorization {
    Authorized,
    Unauthorized,
}
// This takes 2 parameters for the notes.
let component = RpoFalcon512ProcedureAcl::new(
    pub_key,
    vec![],
    NoteAuthorization::Authorized,
    NoteAuthorization::Unauthorized
);
```

But it only replaces bools, but still doesn't use the type system to prevent an accidental mix-up.

**Option 2**

```rust
enum NoteAuthorization {
    AllowUnauthorizedInputNotes,
    AllowUnauthorizedOutputNotes,
    AllowUnauthorizedInputAndOutputNotes,
}
// This takes just 1 parameter for the notes now.
let component = RpoFalcon512ProcedureAcl::new(
    pub_key,
    vec![],
    NoteAuthorization::AllowUnauthorizedOutputNotes
);
```

This is a bit better, but enums are best for cases where the variants are mutually exclusive and here we'd like to combine them.

So bitflags (see this comment: https://github.com/0xMiden/miden-base/pull/1602#discussion_r2216055755) seem most appropriate given that they can be easily combined when necessary and otherwise look and feel like enums.

PhilippGackstatter replied:
The simplest option may be to just have:

```rust
struct AuthRpoFalcon512AclConfig {
    pub allow_unauthorized_output_notes: bool,
    pub allow_unauthorized_input_notes: bool,
}
```

and then `AuthRpoFalcon512Acl::new` would look like:

```rust
let component = AuthRpoFalcon512Acl::new(
    public_key,
    vec![],
    AuthRpoFalcon512AclConfig {
        allow_unauthorized_output_notes: true,
        allow_unauthorized_input_notes: false,
    },
)
```

which simply achieves that the names of the fields are inlined and it's easy to match the field name with the bool.

PhilippGackstatter commented on crates/miden-testing/tests/auth/rpo_falcon_procedure_acl.rs:
```diff
@@ -165,3 +167,202 @@ fn test_rpo_falcon_procedure_acl() -> anyhow::Result<()> {
+    let mut get_item_proc_root = None;
+    let mut set_item_proc_root = None;
+
+    for module in component.library().module_infos() {
```
Nit: I'd change this to use `component.library().get_procedure_root_by_name()` which should be more concise.

PhilippGackstatter commented on crates/miden-testing/tests/auth/rpo_falcon_procedure_acl.rs:
```diff
@@ -165,3 +167,202 @@ fn test_rpo_falcon_procedure_acl() -> anyhow::Result<()> {
+    let set_item_proc_root = set_item_proc_root.expect("set_item procedure should exist");
+    let auth_trigger_procedures = vec![get_item_proc_root, set_item_proc_root];
+
+    let (auth_component, _authenticator) = Auth::ProcedureAcl {
+        auth_trigger_procedures: auth_trigger_procedures.clone(),
+        allow_unauthorized_output_notes: true,
+        allow_unauthorized_input_notes: true,
+    }
+    .build_component();
+
+    let account = AccountBuilder::new([0; 32])
+        .with_auth_component(auth_component)
+        .with_component(component)
+        .account_type(AccountType::RegularAccountUpdatableCode)
+        .storage_mode(AccountStorageMode::Public)
+        .build_existing()?;
```
Nit: We could let `MockChainBuilder::add_account_from_builder` handle some of the setup here to make this test a bit shorter, but this also works.

PhilippGackstatter commented on crates/miden-testing/tests/auth/rpo_falcon_procedure_acl.rs:
```diff
@@ -165,3 +167,202 @@ fn test_rpo_falcon_procedure_acl() -> anyhow::Result<()> {
+    let mut get_item_proc_root = None;
+    let mut set_item_proc_root = None;
+
+    for module in component.library().module_infos() {
```
Similar comment here regarding use of `get_procedure_root_by_name`.

PhilippGackstatter commented on crates/miden-testing/tests/auth/rpo_falcon_procedure_acl.rs:
```diff
@@ -165,3 +167,202 @@ fn test_rpo_falcon_procedure_acl() -> anyhow::Result<()> {
+    Ok(())
+}
+
+#[test]
+fn test_rpo_falcon_procedure_acl_with_disallow_unauthorized_input_notes() -> anyhow::Result<()> {
```
Do you think we could try to share some of the test setup between this and the previous test? So basically create a single `setup_rpo_falcon_procedure_acl_test` function that create the account, mock chain, etc.
It's nice to have short tests and deduplicated test setups so it doesn't take up too much time to understand them.

Alternatively, we could try to share a setup of a test by setting up an account with:

```rust
let (auth_component, _authenticator) = Auth::ProcedureAcl {
    auth_trigger_procedures: auth_trigger_procedures.clone(),
    allow_unauthorized_output_notes: false,
    allow_unauthorized_input_notes: true,
}
.build_component();
```

And then testing that:
- transaction 1: consuming a P2ID note without authenticator works.
- transaction 2: consuming a spawn note (`create_spawn_note`) fails (since it creates an output note).

That way we could share the entire test setup.

Just something to look into and feel free to go with whatever is easier.

PhilippGackstatter reviewed (commented):
Looks good to me!

Left a lot of nits, but the core of it is great :muscle:.

partylikeits1983 commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -16,19 +16,27 @@ const.NUM_AUTH_TRIGGER_PROCS_SLOT=1
 #! Inputs:  [pad(16)]
 #! Outputs: [pad(16)]
-export.auth__tx_rpo_falcon512_procedure_acl
+export.auth__tx_rpo_falcon512_procedure_acl.2
```
Modified to drop `AUTH_ARGS`

partylikeits1983 commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -62,10 +70,44 @@ export.auth__tx_rpo_falcon512_procedure_acl
         # => [should_continue, was_any_called', i-1, pad(16)]
     end
-
     # => [was_any_called, i-1, pad(16)]
```
I renamed `was_any_called` to `require_acl_auth`. I think this makes more sense because I have `require_input_note_auth` and `require_output_note_auth`, then the result of all `or` operations is `require_auth`.

What do you think?

partylikeits1983 commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -16,19 +16,27 @@ const.NUM_AUTH_TRIGGER_PROCS_SLOT=1
+    push.NUM_AUTH_TRIGGER_PROCS_SLOT exec.account::get_item
+    # => [num_auth_trigger_procs, allow_unauthorized_output_notes, allow_unauthorized_input_notes, 0, pad(16)]
+
+    drop
+    # => [allow_unauthorized_input_notes, allow_unauthorized_output_notes, num_auth_trigger_procs, pad(16)]
```
the stack comment was incorrect, I fixed it.

PhilippGackstatter commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -16,61 +16,115 @@ const.NUM_AUTH_TRIGGER_PROCS_SLOT=1
+    # => [!allow_unauthorized_output_notes, were_output_notes_created, require_acl_auth, pad(16)]
 
-    # If any trigger procedure was called, perform signature verification
+    and
+    # => [require_output_note_auth, require_acl_auth, pad(16)]
+
```
Nit: Should we already `or` these together here instead of later? It would make the stack state a bit easier in the following lines. But if you prefer the current logic, that's also fine.

bobbinth replied:
I would slightly prefer to build the condition incrementally - e.g., by doing:
```
and
# => [require_output_note_auth, require_acl_auth, pad(16)]

or
# => [auth_required, pad(16)]
```

PhilippGackstatter commented on crates/miden-lib/src/account/auth/mod.rs:
```diff
@@ -43,39 +43,75 @@ impl From<AuthRpoFalcon512> for AccountComponent {
 /// - Slot 2(map): A map with trigger procedure roots
 ///
 /// This component supports all account types.
 pub struct AuthRpoFalcon512Acl {
```
It might be good to mention in the docs of this as well the caveats from https://github.com/0xMiden/miden-base/issues/1623, i.e. the fact that `was_procedure_called` will only return true if the procedure in question called into a kernel account API that is only allowed to be called from the account context. (This is quite complex and I wonder if anyone will understand this without knowing kernel internals...).

PhilippGackstatter commented on crates/miden-testing/tests/auth/rpo_falcon_procedure_acl.rs:
```diff
@@ -165,3 +169,138 @@ fn test_rpo_falcon_procedure_acl() -> anyhow::Result<()> {
+
+#[test]
+fn test_rpo_falcon_procedure_acl_with_allow_unauthorized_output_notes() -> anyhow::Result<()> {
+    let (component, auth_trigger_procedures, note) = setup_rpo_falcon_procedure_acl_test()?;
```
Nit: To deduplicate a bit more, we can also make `setup_rpo_falcon_procedure_acl_test` take `allow_unauthorized_output_notes: bool, allow_unauthorized_input_notes: bool` and then move the `AccountBuilder` and `MockChain`-building parts into it as well.

PhilippGackstatter reviewed (approved):
Looks good to me!

Thank you for addressing the comments and adding the test setup function, that helps a lot!

bobbinth commented on crates/miden-lib/src/account/auth/mod.rs:
```diff
@@ -43,39 +43,75 @@ impl From<AuthRpoFalcon512> for AccountComponent {
 
-/// An [`AccountComponent`] implementing a procedure-ACL based RpoFalcon512 signature scheme for
-/// authentication of transactions.
+/// An [`AccountComponent`] implementing a procedure-based Access Control List (ACL) RpoFalcon512
+/// signature scheme for authentication of transactions.
```
nit: I would write this as:
```rust
/// An [`AccountComponent`] implementing a procedure-based Access Control List (ACL) using the
/// RpoFalcon512 signature scheme for authentication of transactions.
```

bobbinth commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -16,61 +16,115 @@ const.NUM_AUTH_TRIGGER_PROCS_SLOT=1
+#! If any of these conditions are true, standard RpoFalcon512 signature verification is performed.
+#! Otherwise, only the nonce is incremented.
+#!
+#! Inputs:  [AUTH_ARGS, pad(16)]
```
nit: technically this should be `[AUTH_ARGS, pad(12)]` because this procedure is expected to be invoked using the `call` instruction which cannot access more than 16 elements of the stack.

Not that `dropw` will still result in `[pad(16)]` because zeros will be shifted in from the right.

bobbinth commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -16,61 +16,115 @@ const.NUM_AUTH_TRIGGER_PROCS_SLOT=1
+
     # Get the number of trigger procedures
-    push.NUM_AUTH_TRIGGER_PROCS_SLOT exec.account::get_item drop drop drop
+    push.NUM_AUTH_TRIGGER_PROCS_SLOT exec.account::get_item
```
We should rename this constant as I mentioned in https://github.com/0xMiden/miden-base/pull/1602#discussion_r2214558312.

partylikeits1983 replied:
Renamed the constant.

bobbinth commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -16,61 +16,115 @@ const.NUM_AUTH_TRIGGER_PROCS_SLOT=1
     if.true
         exec.::miden::contracts::auth::basic::auth__tx_rpo_falcon512
     else
         exec.account::incr_nonce
```
Question: should we increment the nonce only if the account was updated? (e.g., we could check if account's initial commitment is different from the current commitment).

PhilippGackstatter replied:
I think that makes sense. That way, accounts that use this auth component to allow note anonymization wouldn't have their nonce incremented as long as the account state isn't touched.

bobbinth commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -16,61 +16,115 @@ const.NUM_AUTH_TRIGGER_PROCS_SLOT=1
+
+    # If authentication is required (either note-based or procedure-based), perform signature verification
     if.true
         exec.::miden::contracts::auth::basic::auth__tx_rpo_falcon512
```
Rather than calling `auth__tx_rpo_falcon512` (which is meant to be invoked via `call` instruction), it may be better to have some helper functions which are shared between that and this auth component.

PhilippGackstatter replied:
Agreed, though it may make sense to do this after https://github.com/0xMiden/miden-base/pull/1624 is merged, since it touches a lot of the falcon code.

bobbinth replied:
Agreed. Maybe in #1624 we can already split out some things into helper procedures - e.g.,

- `prepare_tx_summary` - puts the 4 words of the tx summary onto the stack.
- `hash_tx_summary` hashes summary.

PhilippGackstatter replied:
That's pretty much what I ended up with (`create_tx_summary` and `hash_tx_summary`) and they are made public in https://github.com/0xMiden/miden-base/pull/1634, though if it's time-critical that can be easily done as part of a separate PR as well.

bobbinth commented on crates/miden-lib/src/account/faucets/mod.rs:
```diff
@@ -246,7 +247,7 @@ pub fn create_basic_fungible_faucet(
     let auth_component: AuthRpoFalcon512Acl = match auth_scheme {
         AuthScheme::RpoFalcon512 { pub_key } => {
-            AuthRpoFalcon512Acl::new(pub_key, vec![distribute_proc_root])
+            AuthRpoFalcon512Acl::new(pub_key, vec![distribute_proc_root], false, false)
```
Should we not set `allow_unauthorized_input_notes = true` here? I think for the faucet we wanted to enable anyone to call `burn` procedure which would require consuming an input note.

bobbinth reviewed (commented):
Thank you! Looks good! I reviewed most non-test code and left some comments inline. The main one is about potentially not calling one auth component from another.

PhilippGackstatter commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -10,67 +10,120 @@ const.FALCON_SIG_TO_STACK=131087
         exec.account::incr_nonce
     end
+    # => [pad(16)]
 
-    drop
+    repeat.4
+        dropw
+    end
+    # => []
```
The expected outputs of this procedure are `pad(16)` since it is `call`ed, so I think we don't actually have to drop anything here. This would just shift in 16 zeroes from the right, also producing `pad(16)`.

PhilippGackstatter reviewed (approved):
Looks good to me!

PhilippGackstatter commented on crates/miden-lib/src/account/auth/mod.rs:
```diff
@@ -170,8 +282,170 @@ mod tests {
+    }
+
+    #[test]
+    fn test_rpo_falcon_512_procedure_acl_with_allow_unauthorized_output_notes() {
```
If you feel like we need more tests, could you open an issue to describe what needs to be tested? Thanks!

partylikeits1983 commented:
@mmagician I refactored the tests for the `rpo_falcon_512_procedure_acl` component so that there is minimal duplication on setup of the tests. 

@PhilippGackstatter Also addressed the masm comments you had regarding `dropw padw` at the beginning of the execution of the component and `repeat.4 dropw end` at the end of the component.

PhilippGackstatter commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -4,67 +4,129 @@ use.miden::tx
+    or
+    # => [auth_required, pad(16)]
+
+    # ------ Check if initial account commitment differs from current commitment ------
+
+    exec.account::get_initial_commitment
+    # => [INITIAL_COMMITMENT, auth_required, pad(16)]
+
+    exec.account::compute_current_commitment
+    # => [CURRENT_COMMITMENT, INITIAL_COMMITMENT, auth_required, pad(16)]
+
+    eqw not movdn.8 dropw dropw
+    # => [has_account_state_changed, auth_required, pad(16)]
+
+    or
+    # => [auth_required, pad(16)]
```
I think doing it this way would sidestep the use of trigger procedures. For example, in the basic fungible faucet, we want to allow `burn` to be called by anyone (not require auth) but `mint` only to be called by the owner (require auth). With this approach, `burn` would cause `has_account_state_changed = true` and so cause us to enter the auth required branch, but that's not what we want.

I think what @bobbinth was referring to, and how I understood it, was doing this in the else branch below. So if `auth_required` as computed on line 109 is false, we enter the else branch and, currently, always increment the nonce. Here, we'd like to change this to only increment the nonce if a state change has actually occurred, and not increment the nonce if no state change occured. Does that make sense?

partylikeits1983 replied:
Ah that makes sense. Will fix.

PhilippGackstatter replied:
Looks good! I think you'll have to change or remove one of the tests as a result of this change.

mmagician commented on crates/miden-lib/src/account/auth/mod.rs:
```diff
@@ -218,249 +218,152 @@ mod tests {
         // TODO refactor to fetch procedure digests by name after
         // https://github.com/0xMiden/miden-base/pull/1532
-        let auth_trigger_procedures: Vec<Word> = basic_wallet_library()
+        let procedures: Vec<Word> = basic_wallet_library()
```
maybe for a follow-up PR: We have `procedure_digests()` on a `WellKnownComponent` but they're not exposed. This would simplify the following code to:
```
let procedures = WellKnownComponent::BasicWallet::procedure_digests().collect();
```

partylikeits1983 replied:
Will implement this in this PR, its pretty straightforward I think

partylikeits1983 replied:
Fixed!

partylikeits1983 commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -4,67 +4,128 @@ use.miden::tx
+    # => [INPUT_NOTES_COMMITMENT, auth_required, pad(16)]
+
+    padw eqw not movdn.8 dropw dropw
+    # => [were_input_notes_consumed, auth_required, pad(16)]
```
Awesome, this actually saves around ~4 cycles

partylikeits1983 commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -4,67 +4,128 @@ use.miden::tx
+    # => [OUTPUT_NOTES_COMMITMENT, require_acl_auth, pad(16)]
+
+    padw eqw not movdn.8 dropw dropw
+    # => [were_output_notes_created, require_acl_auth, pad(16)]
```
Updated to use `tx::get_num_output_notes`.

bobbinth commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -76,10 +76,10 @@ export.auth__tx_rpo_falcon512_procedure_acl.2
+    # => [num_output_notes, require_acl_auth, pad(16)]
 
-    padw eqw not movdn.8 dropw dropw
+    push.0 eq not
```
nit: this could be just `neq.0`, right?

bobbinth commented on crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm:
```diff
@@ -93,10 +93,10 @@ export.auth__tx_rpo_falcon512_procedure_acl.2
     # => [INPUT_NOTES_COMMITMENT, auth_required, pad(16)]
 
-    padw eqw not movdn.8 dropw dropw
+    push.0 eq not
```
nit: this could be just `neq.0`, right?

bobbinth reviewed (approved):
Looks good! Thank you! I left just two tiny nits inline.
